mploop.1
.Dd June 06, 2026
.Os Linux
.Dt MPLOOP 1 LOCAL
.Sh NAME
.Nm mploop
.Nd the queue-based command line music player
.Sh SYNOPSIS
.Nm mploop
.Sh DESCRIPTION
.Nm mploop
is a tool to play music from a queue using a command line interface to manage
the queue.
.Nm mploop
is not a daemon, it uses the terminal. It is intended to be started in a
.Xr screen 1
or a
.Xr tmux 1
window.
.Pp
Although
.Nm mploop
is not a daemon, out of the existing music players it is most similar with
.Xr mpd 1 .
However, unlike
.Xr mpd 1 ,
mploop does not want to scan your entire music library. Instead, arbitrary
files can be added to the queue from any directory.
.Sh "QUEUE MANAGEMENT"
The queue is managed by the following commands:
.Bl -hang -offset 5 -width "012345678901234"
.It Xr mpq 1
Display queue or enqueue new songs.
.It Xr mprm 1
Remove songs from queue.
.It Xr mpclear 1
Clear the entire queue
.It Xr mpshuffle 1
Shuffle the entire queue
.It Xr vimp 1
Perform more complex queue management by a text editor
.El
.Sh "COMMAND-LINE CONTROL"
The player can be controlled from the command line using these commands:
.Bl -hang -offset 5 -width "012345678901234"
.It Xr mpprev 1
Jump to previous song or N songs backwards
.It Xr mpnext 1
Jump to next song or N songs forwards
.It Xr mpplaypause 1
Play or pause the playback
.It Xr mprewind 1
Rewind the current song
.It Xr mpseek 1
Perform more complex seeks backwards/forwards in the current song
.El
.Sh "INTERACTIVE CONTROL"
To play music,
.Nm mploop
uses either
.Xr mplayer 1
or the included player mploopplayer. Both share the same basic shortcuts.
.Pp
.Sy "keyboard control"
.Pp
.Bl -hang -offset 5 -width "01234567890123456789"
.It "LEFT and RIGHT"
Seek backward/\:forward 10 seconds.
.It "UP and DOWN"
Seek forward/\:backward 1 minute.
.It "PGUP and PGDWN"
Seek forward/\:backward 10 minutes.
.It "q / ENTER"
Go forward in the playlist.
.It "p / SPACE"
Pause (pressing again unpauses).
.It "/ and *"
Decrease/increase volume.
.It "9 and 0"
Decrease/increase volume.
.El
.Sh "FILE FORMAT SUPPORT"
.Nm mploop
can play files using the included mploopplayer or by using
.Xr mplayer 1 .
If the included mploopplayer is used, then
.Nm mploop
can play back all formats supported by libavformat and libavcodec.
.Pp
ReplayGain is supported for Ogg Vorbis, Opus, FLAC, MP3 and MP4 (AAC).
Also comment tags are supported for the same formats. The comment tags are
automatically converted to Vorbis comments and then displayed.
.Sh "AUDIO DRIVER SUPPORT"
.Nm mploop
can output audio using the included mploopplayer or by using
.Xr mplayer 1 .
If the included mploopplayer is used, then
.Nm mploop
plays back audio using SDL2 so it supports all audio drivers that the SDL2
library supports.
.Sh "EXIT STATUS"
.Nm mploop
does not exit except when interrupted, since it waits for music to arrive
into the queue to play it. If there's an error in mploop itself,
.Nm mploop
returns with a nonzero exit status. However, if there's an error in playing
music,
.Nm mploop
skips the song where the error occurred and continues playing music.
.Sh ENVIRONMENT
.Nm mploop
uses the following environment variables:
.Pp
.Bl -column HOME -compact
.It Ev HOME Ta the home directory of the user
.El
.Sh "SEE ALSO"
.Xr mpclear 1 ,
.Xr mpnext 1 ,
.Xr mpnp 1 ,
.Xr mpplaypause 1 ,
.Xr mpprev 1 ,
.Xr mpq 1 ,
.Xr mprewind 1 ,
.Xr mprm 1 ,
.Xr mpseek 1 ,
.Xr mpshuffle 1 ,
.Xr vimp 1
.Sh HISTORY
.Xr mploop 1
was created in September to November 2023. It has been maintained since.
.Sh DISTRIBUTION
.Xr mploop 1
is available at https://github.com/jmtilli/mploop
.Sh BUGS
.Bl -bullet
.It
.Nm mploop
does not support gapless playback.
.El
.Pp
Also, some utilities like
.Xr mpseek 1
and
.Xr mprewind 1
have bugs.
.Sh AUTHOR
Juha-Matti Tilli (juha-matti.tilli@iki.fi)
.Sh COPYRIGHT
Copyright \(co 2023-4,2026 Juha-Matti Tilli
.Pp
Licensed under the MIT license.


mploop.py
#!/usr/bin/env python3
from __future__ import print_function
from __future__ import division
import os
import sys
import time
import socket
import subprocess
import termios
import select
import libmploop
import errno

libmploop.touch()

# Mandatory tools to run:
# - mplayer
# - file
# Recommended tools to run:
# - id3v2
# - id3tool
# - mp3gain
# - metaflac
# - vorbiscomment
# - opusinfo or opustags
# - AtomicParsley

offset2 = 6.0
mploopplayer_extraoffset = 0.0

mainlck = libmploop.MainLock()

adaptive_sleep = libmploop.AdaptiveSleep()

toclear = True

try:
    while True:
        if os.stat(libmploop.dbexpanded).st_size == 0:
            if toclear:
                with open(libmploop.npexpanded, "w") as f:
                    f.write('')
                toclear = False
            adaptive_sleep.sleep()
            continue
        with libmploop.DbLock() as lck:
            with open(libmploop.dbexpanded, 'r') as f:
                ln = f.readline()
                if ln == '':
                    if toclear:
                        with open(libmploop.npexpanded, "w") as f:
                            f.write('')
                        toclear = False
                    adaptive_sleep.sleep()
                    continue
                adaptive_sleep.seen()
                if ln and ln[-1] == '\n':
                    ln = ln[:-1]
                rawln = ln
                ln = libmploop.unescape(ln)
                rest = ''.join(f.readlines())
            with open(libmploop.dbexpanded, "w") as f:
                f.write(rest)
            with open(libmploop.pastexpanded, 'r') as f:
                allpast = rawln + '\n' + ''.join(f.readlines())
            with open(libmploop.pastexpanded, 'w') as f:
                f.write(allpast)
        if ln and ln[0] == '/':
            hasgain,gain,comments = libmploop.get_gain(ln)
        else:
            hasgain = False
            gain = 0.0
            comments = []
        if libmploop.clear_stdin():
            print("")
        if not libmploop.mploopplayer:
            print(80*"=")
            print("Applying gain:", gain-offset2)
            print("File:", ln)
        for comment in comments:
            k = comment[0]
            pretty = k
            if k == 'TRACKNUMBER':
                pretty = 'Track number:'
            elif k == 'COPYRIGHT':
                pretty = 'Copyright'
            elif k == '':
                pretty = 'Comment:'
            else:
                pretty = k[0:1].upper() + k[1:].lower() + ':'
            v = comment[1]
            print(pretty + ' ' + v)
        if not libmploop.mploopplayer:
            with open(os.path.expanduser('~') + '/.mploop/np.txt', "w") as f:
                if hasattr(rawln, "decode"): # Python 2
                    f.write(("FILE=" + rawln.decode("utf-8") + '\n' + '\n'.join(c[0] + "=" + c[1] for c in comments) + '\n').encode("utf-8"))
                else: # Python 3
                    f.write("FILE=" + rawln + '\n' + '\n'.join(c[0] + "=" + c[1] for c in comments) + '\n')
        toclear=True
        if not libmploop.mploopplayer:
            print(80*"-")
        if libmploop.mploopplayer:
            urlargs = []
            npf = os.path.expanduser('~') + '/.mploop/np.txt'
            if ln and ln[0] != '/':
                urlargs = ["-u"]
            if hasgain:
                proc = subprocess.Popen([libmploop.mploopplayer] + urlargs + ["-s", libmploop.sockexpanded, "-W", npf, "-g", str(gain-offset2-mploopplayer_extraoffset), "--", ln])
            else:
                proc = subprocess.Popen([libmploop.mploopplayer] + urlargs + ["-s", libmploop.sockexpanded, "-G", "-W", npf, "-g", str(0.0-offset2-mploopplayer_extraoffset), "--", ln])
            with open(libmploop.mploopplayerpidexpanded, 'w') as f:
                f.write(str(proc.pid) + '\n')
            proc.wait()
        else:
            proc = subprocess.Popen(["mplayer", "-novideo", "-nolirc", "-msglevel", "all=0:statusline=5:cplayer=5", "-af", "volume=" + str(gain-offset2) + ":1", "--", ln], stdin=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0)
            with open(libmploop.mplayerpidexpanded, 'w') as f:
                f.write(str(proc.pid) + '\n')
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            try:
                os.unlink(libmploop.mplayersockexpanded)
            except OSError as e:
                if e.errno != errno.ENOENT:
                    raise
                pass
            sock.bind(libmploop.mplayersockexpanded)
            sock.listen(16)
            fds = set([])
            p = select.poll()
            p.register(sys.stdin, select.POLLIN)
            p.register(proc.stderr, select.POLLIN)
            p.register(sock, select.POLLIN)
            eof = False
            old_settings = termios.tcgetattr(sys.stdin)
            new_settings = termios.tcgetattr(sys.stdin)
            new_settings[3] = new_settings[3] & ~termios.ICANON & ~termios.ECHO
            termios.tcsetattr(sys.stdin, termios.TCSANOW, new_settings)
            socks = {}
            try:
                while not eof:
                    for fd, revent in p.poll():
                        if fd == sock.fileno() and (revent & select.POLLIN):
                            newfd, cliaddr = sock.accept()
                            fds.add(newfd.fileno())
                            p.register(newfd.fileno(), select.POLLIN)
                            socks[newfd.fileno()] = newfd
                        elif fd == sys.stdin.fileno() and (revent & select.POLLIN):
                            buf = os.read(sys.stdin.fileno(), 4096)
                            proc.stdin.write(buf)
                            proc.stdin.flush()
                        elif fd == proc.stderr.fileno() and (revent & (select.POLLIN | select.POLLHUP)):
                            buf = os.read(proc.stderr.fileno(), 4096)
                            if buf == b'':
                                eof = True
                                continue
                            sys.stderr.write(buf)
                            sys.stderr.flush()
                        elif fd != sock.fileno() and fd != sys.stdin.fileno() and fd != proc.stderr.fileno() and (revent & (select.POLLIN | select.POLLHUP)):
                            buf = os.read(fd, 4096)
                            if buf == b'':
                                p.unregister(fd)
                                socks[fd].close()
                                fds.remove(fd)
                                continue
                            proc.stdin.write(buf)
                            proc.stdin.flush()
            finally:
                termios.tcsetattr(sys.stdin, termios.TCSANOW, old_settings)
            for newfd in fds:
                socks[newfd].close()
            sock.close()
            proc.wait()
        print("")
except KeyboardInterrupt:
    with open(libmploop.npexpanded, "w") as f:
        f.write('')
    print("")
